rtl/eclrtl/rtlfield.cpp

```diff
@@ -4213,6 +4213,99 @@ unsigned ECLRTL_API countFields(const RtlFieldInfo * const * fields)
     return cnt;
 }
 
+//-------------------------------------------------------------------------------------------------------------------
+
+static const unsigned sameTypeMask = RFTMkind|RFTMebcdic;
+size32_t translateScalar(ARowBuilder &builder, size32_t offset, const RtlFieldInfo *field, const RtlTypeInfo &destType, const RtlTypeInfo &sourceType, const byte *source)
+{
+    switch(destType.getType())
+    {
+    case type_boolean:
+    case type_int:
+    case type_swapint:
+    case type_packedint:
+    case type_filepos:
+        offset = destType.buildInt(builder, offset, field, sourceType.getInt(source));
+        break;
+    case type_real:
+        offset = destType.buildReal(builder, offset, field, sourceType.getReal(source));
+        break;
+    case type_data:
+    case type_string:
+        //Special case if source and destination types are identical to avoid cloning strings
+        if ((destType.fieldType & sameTypeMask) == (sourceType.fieldType & sameTypeMask))
```

Comment:
This function moved - but this particular part of the code is new - to avoid cloning strings when resizing them.  It will also potentially optimize some cases of projecting fields whether the change is the length that are not already special cased.